src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionCommitManager.cs

```diff
@@ -0,0 +1,170 @@
+            }
+
+            return CommitResultUnhandled;
+        }
+
+        /// <summary>
+        /// This method needs to support custom procesing of commit characters to be on par with the old completion implementation.
+        /// </summary>
+        private bool IsCommitCharacter(char typeChar, ImmutableArray<CharacterSetModificationRule> rules)
+        {
+            // Tab, Enter and Null (call invoke commit) are always a commit character
+            if (typeChar == '\t' || typeChar == '\n' || typeChar == '\0')
+            {
+                return true;
+            }
+
+            foreach (var rule in rules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return true;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return false;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(typeChar);
+                }
+            }
+
+            return s_commitChars.Contains(typeChar);
+        }
+
+        private EditorCompletion.CommitBehavior CustomCommit(
+            Document document,
+            IFeaturesCustomCommitCompletionProvider provider,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            ITrackingSpan applicableSpan,
+            char commitCharacter,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                return EditorCompletion.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                return EditorCompletion.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
+            {
+                var currentViewSpan = applicableSpan.GetSpan(applicableSpan.TextBuffer.CurrentSnapshot);
+                var spans = view.BufferGraph.MapDownToBuffer(currentViewSpan, SpanTrackingMode.EdgeInclusive, subjectBuffer);
+
+                // There can be no spans in case of projections but must be less than 2 spans.
+                // 1 span is a regular case.
+                // 0 spans means the a span is in anoter projection.
+                // We do not expect more than 1 span, and do not support this case.
+                System.Diagnostics.Debug.Assert(spans.Count < 2);
```

Comment:
if we don't support this case, tha'ts different than asserting this case can't happen.